### PR TITLE
xpath BUGFIX added max number of nested exprs

### PR DESCRIPTION
**Pull request by lePici**
The change prevents stack-overflow error.
This PR fixes issue 34052 from oss-fuzz.

### CODE REVIEW

**Review by michalvasko — CHANGES_REQUESTED** [review]

File: src/common.h [hunk 1]
```diff
@@ -274,6 +274,7 @@ size_t LY_VCODE_INSTREXP_len(const char *str);
 #define LY_VCODE_XP_INOP_1      LYVE_XPATH, "Cannot apply XPath operation %s on %s."
 #define LY_VCODE_XP_INOP_2      LYVE_XPATH, "Cannot apply XPath operation %s on %s and %s."
 #define LY_VCODE_XP_INMOD       LYVE_XPATH, "Unknown/non-implemented module \"%.*s\"."
+#define LY_VCODE_XP_DEPTH       LYVE_XPATH, "The maximum number of round/square bracket nestings has been exceeded."
```
[the remark this answers is not in the document]
Comment (lePici) [reply src/common.h:277]: ok

File: src/tree_schema_internal.h [hunk 2]
```diff
@@ -39,6 +39,7 @@ struct lys_glob_unres;
  * Designed to protect against corrupted input that causes a stack-overflow error.
  * For yang language and json format, the block is bounded by "{ }".
  * For the xml format, the opening and closing element tag is considered as the block.
+ * For the XPath format, the "[ ]" or "( )" is considered as the block.
```
[the remark this answers is not in the document]
Comment (lePici) [reply src/tree_schema_internal.h:42]: ok

File: src/xpath.c [hunk 3]
```diff
@@ -1943,23 +1943,27 @@ exp_repeat_push(struct lyxp_expr *exp, uint16_t tok_idx, uint16_t repeat_op_idx)
...
+    LY_CHECK_ERR_RET(depth > LY_MAX_BLOCK_DEPTH, LOGVAL(ctx, LY_VCODE_XP_DEPTH), LY_EINVAL);
 
-    rc = reparse_or_expr(ctx, exp, tok_idx);
+    rc = reparse_or_expr(ctx, exp, tok_idx, depth);
     LY_CHECK_RET(rc);
 
     rc = lyxp_check_token(ctx, exp, *tok_idx, LYXP_TOKEN_BRACK2);
     LY_CHECK_RET(rc);
     ++(*tok_idx);
+    --depth;
```
[the remark this answers is not in the document]
Comment (lePici) [reply src/xpath.c:1968]: ok

File: src/xpath.c [hunk 4]
```diff
@@ -1943,23 +1943,27 @@ exp_repeat_push(struct lyxp_expr *exp, uint16_t tok_idx, uint16_t repeat_op_idx)
...
     LY_ERR rc;
 
     rc = lyxp_check_token(ctx, exp, *tok_idx, LYXP_TOKEN_BRACK1);
     LY_CHECK_RET(rc);
     ++(*tok_idx);
+    ++depth;
+    LY_CHECK_ERR_RET(depth > LY_MAX_BLOCK_DEPTH, LOGVAL(ctx, LY_VCODE_XP_DEPTH), LY_EINVAL);
 
-    rc = reparse_or_expr(ctx, exp, tok_idx);
+    rc = reparse_or_expr(ctx, exp, tok_idx, depth);
```
[the remark this answers is not in the document]
Comment (lePici) [reply src/xpath.c:1962]: ok

File: src/xpath.c [hunk 5]
```diff
@@ -7413,7 +7411,7 @@ eval_relative_location_path(const struct lyxp_expr *exp, uint16_t *tok_idx, ly_b
             break;
 
         default:
-            LOGINT_RET(set ? set->ctx : NULL);
+            LOGINT_RET(options & LYXP_SKIP_EXPR ? NULL : set->ctx);
```
[the remark this answers is not in the document]
Comment (lePici) [reply src/xpath.c:7416]: ok

**Review by michalvasko — APPROVED** [review]
Adding the skip-expr flag is not needed anymore but it is a good change on its own so it can be merged.

**Pull request merged**